File: MgDev/Oem/dbxml-2.3.10/dbxml/examples/java/gettingStarted/modifyDocument.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2006 Oracle.  All rights reserved.
//
// $Id: modifyDocument.java,v 1.14.2.2 2007/01/30 15:34:40 gmf Exp $
//

package dbxml.gettingStarted;

import java.io.*;

import com.sleepycat.dbxml.*;

class modifyDocument
{

    private static String theContainer = "namespaceExampleData.dbxml";
    private static String collection = "collection('namespaceExampleData.dbxml')";
    private static void usage() {
	String usageMessage = "\nThis program modifies documents found in a DBXML container.\n";
	usageMessage += "You should run exampleLoadContainer before running this example.\n";
	usageMessage += "You are only required to pass this command the path location of the\n";
	usageMessage += "database environment that you specified when you loaded the examples data:\n\n";
	usageMessage += "\t-h <dbenv directory>\n";
	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.modifyDocument -h examplesEnvironment\n";
	System.out.println(usageMessage);
	System.exit( -1 );
    }

    // display documents matching the query
    private static void dumpDocuments(XmlResults results)
	throws Throwable {
	System.out.println("Found " + results.size() + " documents ");

	results.reset();
	XmlValue value = results.next();
	while(value != null) {
            XmlDocument document = value.asDocument();
            String docString =document.getContentAsString();
            System.out.println("Document content: \n" + docString);
	    document.delete();
	    value.delete();
            value = results.next();
	}
    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {
	File path2DbEnv = null;
	for(int i = 0; i < args.length; ++i) {
            if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File( args[++i] );
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	myDbEnv env = null;
	XmlContainer openedContainer = null;
	XmlTransaction txn = null;
	try {
	    // Open an environment
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    // open a transactional container
	    XmlContainerConfig config = new XmlContainerConfig();
	    config.setTransactional(true);
	    openedContainer =
		theMgr.openContainer(theContainer, config);

	    // Start atransaction
	    txn = theMgr.createTransaction();

	    //Create a context and declare the namespaces
	    XmlQueryContext context =
		theMgr.createQueryContext(XmlQueryContext.LiveValues,
					  XmlQueryContext.Eager);

	    context.setNamespace( "fruits", "http://groceryItem.dbxml/fruits");
	    context.setNamespace( "vegetables",
				  "http://groceryItem.dbxml/vegetables");
	    context.setNamespace( "desserts", "http://groceryItem.dbxml/desserts");

	    String query = collection +
		"/fruits:item/product[. = 'Zapote Blanco']";
	    XmlQueryExpression expression = theMgr.prepare(txn,
							   query, context);

	    System.out.println("Updating document for the expression: '" +
			       query + "' ");
	    System.out.println("Return to continue: ");
	    System.in.read();

	    //Print the document(s) to be updated -- those that describe
	    // Zapote Blanco (a fruit). The document strings are only being
	    // printed to show before/after results.
	    // Most modification programs would not perform the additional queries.
	    XmlResults results = expression.execute(txn, context, null);
	    dumpDocuments(results);
	    results.reset();
	    System.out.println("About to update the document(s) above.\n" +
			       "Look for a new element after the target " +
			       "element, named 'description' ...\n" +
			       "Return to continue: ");
	    System.in.read();

	    //The modification is a new element in the target node, called
	    // "descripton, which goes immediately after the <product> element.
	    // if this program is run more than once, and committed, additional
	    // identical elements are added.  It is easy to modify this program
	    // to change the modification.
	    XmlModify modify = theMgr.createModify();
	    modify.addInsertAfterStep(theMgr.prepare(txn, ".", context),
				      XmlModify.Element,
				      "description", "very nice");

	    XmlUpdateContext uc = theMgr.createUpdateContext();
	    results.reset();
	    long numMod = modify.execute(txn, results, context, uc);

	    System.out.println("Peformed " + numMod + " modification operations");
	    dumpDocuments(results);

	    System.out.println("If committed, this program will add a new element each time it is run.\n" + "Press 'c' to commit this change:");
	    int c = System.in.read();
	    if (c == 'c' || c =='C')
		txn.commit();
	    else
		txn.abort();
	    txn.delete();
	    expression.delete();
	    modify.delete();
	    results.delete();
	    context.delete();
	    uc.delete();

	} catch(Exception e) {
	    System.err.println("Error performing document modification against "
			       + theContainer);
	    System.err.println("   Message: " + e.getMessage());
	    // In the event of an error, we abort the operation
	    // The database is left in the same state as it was in before
	    // we started this operation.
	    if ( txn != null ) {
		txn.abort();
		txn.delete();
	    }
	    throw e;
	}
	finally {
	    cleanup(env, openedContainer);
	}
    } //End main
}

File: MgDev/Oem/dbxml-2.3.10/dbxml/src/perl/examples/gettingStarted/modifyDocument.pl
use strict;
use warnings;
use Sleepycat::DbXml 'simple';


sub usage()
{
    warn  <<EOM;
This program modifies documents found in a DBXML container. You should
pre-load the container using loadExamplesData.[sh|cmd] before running this
example. You are only required to pass this command the path location of the
database environment that you specified when you pre-loaded the examples
data:

    -h <dbenv directory>
EOM
        exit( -1 );
}

#  print out the string values of matching documents
sub dumpDocuments($$)
{
    my $mgr = shift ;
    my $results = shift ;

    # //////////////////////////////////////////////////////////////////////
    # ////  Updates a document stored in a DBXML container.          ///////
    # //////////////////////////////////////////////////////////////////////
	
    # /// some defensive code eliminated for clarity //
	
	eval {
        $results->reset();
		print "Found " . $results->size() . " documents.\n\n";
		my $doc = $mgr->createDocument;
		while( $results->next($doc) ) {
			print "Document content: $doc\n";
		}
	};
    if (my $e = catch std::exception)
    {
        warn "Document query failed.\n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
        warn "Document query failed.\n";
        warn $@;
        exit( -1 );
    }
}

my $path2DbEnv;
my $theContainer = "namespaceExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
  if ! $path2DbEnv ;

# open a container in the db environment
my $env = new DbEnv(0);
$env->set_cachesize(0, 64 * 1024, 1);
$env->open($path2DbEnv,
            Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
my $theMgr = new XmlManager($env);

my $containerTxn = $theMgr->createTransaction();
my $container = $theMgr->openContainer($containerTxn, $theContainer);
$containerTxn->commit();


# create a context and declare the namespaces
my $context = $theMgr->createQueryContext(XmlQueryContext::LiveValues, XmlQueryContext::Eager);
$context->setNamespace( "fruits", "http://groceryItem.dbxml/fruits");
$context->setNamespace( "vegetables", "http://groceryItem.dbxml/vegetables");
$context->setNamespace( "desserts", "http://groceryItem.dbxml/desserts");
my $query = "collection('" . $container->getName() . "')/fruits:item/product[text() = 'Zapote Blanco']";
eval {
        # myDbEnv and myXmlContainer open with transactions. All subsequent
        # writes to them must also be performed inside a transaction.
        # Get a transaction
		my $txn = $theMgr->createTransaction();
		
		my $expression = $theMgr->prepare($txn, $query, $context);
		
		print "Updating document for the expression: '$query'\n";
		print "Return to continue: ";
		<>;
		
        #  print the document(s) to be updated -- those that describe Zapote Blanco (a fruit)
        #  The document strings are only being printed to show before/after results.
        #  Most modification programs would not perform the additional queries.
		my $results = $expression->execute($txn, $context);
		dumpDocuments($theMgr, $results);
		
		print "About to update the document(s) above.\nLook for a new element after the target element, named 'description' ...\n";
		print "Return to continue: ";
		<>;
		
        #  the modification is a new element in the target node, called "descripton,
        #  which goes immediately after the <product> element.
        #  if this program is run more than once, and committed, additional
        #  identical elements are added.  It is easy to modify this program
        #  to change the modification.
		my $modify = $theMgr->createModify();
		$modify->addInsertAfterStep($theMgr->prepare($txn, ".", $context), XmlModify::Element, "description", "very nice");
		
		my $uc = $theMgr->createUpdateContext();
		$results->reset();
		my $noOfModifications = $modify->execute($txn, $results, $context, $uc);
		print "$noOfModifications modifications made.\n";
		
        #  print the modified documents
		dumpDocuments($theMgr, $results);
		
		print "If committed, this program will add a new element each time it is run.\n";
		print "Press 'c' to commit this change:";
		my $cmt = <> ;
		if ($cmt =~ /^c/i) {
			$txn->commit();
		}
		else {
			$txn->abort();
		}
};
if (my $e = catch std::exception)
{
    warn "Document modification failed.\n";
    warn $e->what() . "\n";
    exit( -1 );
}
elsif ($@)
{
    warn "Document modification failed.\n";
    warn $@;
    exit( -1 );
}
